pydantic_graph/pydantic_graph/persistence/__init__.py

```diff
@@ -0,0 +1,205 @@
+
+[`Graph.run`][pydantic_graph.graph.Graph.run] returns a list of these steps describing the execution of the graph,
+together with the run return value.
+"""
+
+
+class BaseStatePersistence(ABC, Generic[StateT, RunEndT]):
+    """Abstract base class for storing the state of a graph."""
+
+    @abstractmethod
+    async def snapshot_node(self, state: StateT, next_node: BaseNode[StateT, Any, RunEndT]) -> None:
+        """Snapshot the state of a graph, when the next step is to run a node.
+
+        Args:
+            state: The state of the graph.
+            next_node: The next node to run.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def snapshot_node_if_new(
+        self, snapshot_id: str, state: StateT, next_node: BaseNode[StateT, Any, RunEndT]
+    ) -> None:
+        """Snapshot the state of a graph, if the snapshot ID doesn't already exist in persistence.
+
+        Args:
+            snapshot_id: The ID of the snapshot to check.
+            state: The state of the graph.
+            next_node: The next node to run.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def snapshot_end(self, state: StateT, end: End[RunEndT]) -> None:
+        """Snapshot the state of a graph after a node has run, when the graph has ended.
+
+        Args:
+            state: The state of the graph.
+            end: data from the end of the run.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def record_run(self, snapshot_id: str) -> AbstractAsyncContextManager[None]:
+        """Record the run of the node, or error if the node is already running.
+
+        Args:
+            snapshot_id: The ID of the snapshot to record.
+
+        Raises:
+            GraphNodeRunningError: if the node status it not `'created'` or `'pending'`.
+            LookupError: if the snapshot ID is not found in persistence.
+
+        Returns:
+            An async context manager that records the run of the node.
+
+        In particular this should set:
+
+        - [`NodeSnapshot.status`][pydantic_graph.persistence.NodeSnapshot.status] to `'running'` and
+          [`NodeSnapshot.start_ts`][pydantic_graph.persistence.NodeSnapshot.start_ts] when the run starts.
+        - [`NodeSnapshot.status`][pydantic_graph.persistence.NodeSnapshot.status] to `'success'` or `'error'` and
+          [`NodeSnapshot.duration`][pydantic_graph.persistence.NodeSnapshot.duration] when the run finishes.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def retrieve_next(self) -> NodeSnapshot[StateT, RunEndT] | None:
+        """Retrieve a node snapshot with status `'created`' and set its status to `'pending'`.
+
+        This is used by [`Graph.next_from_persistence`][pydantic_graph.graph.Graph.next_from_persistence]
+        to get the next node to run.
+
+        Returns: The snapshot, or `None` if no snapshot with status `'created`' exists.
```

Comment:
This makes me feel like I'm missing something about how this works. Probably need to discuss synchronously, but I don't understand why some methods accept a `snapshot_id` and others don't.

Reply:
discussed.